src/Geometry.cpp

```diff
@@ -528,48 +516,49 @@ int Geometry::findFSRId(LocalCoords* coords) {
   std::size_t fsr_key_hash = key_hash_function(getFSRKey(coords));
 
   /* If FSR has not been encountered, update FSR maps and vectors */
-  if (_FSR_keys_map.find(fsr_key_hash) == _FSR_keys_map.end()){
-
-    /* Get the cell that contains coords */
-    Cell* cell = findCellContainingCoords(curr);
-    
-    /* Get the lock */
-    omp_set_lock(_num_FSRs_lock);
-
-    /* Recheck to see if FSR has been added to maps after getting the lock */
-    if (_FSR_keys_map.find(fsr_key_hash) != _FSR_keys_map.end())
-      fsr_id = _FSR_keys_map.at(fsr_key_hash)._fsr_id;
+  if (!_FSR_keys_map.contains(fsr_key_hash)){
+
+    /* Try to get a clean copy of the fsr_id, adding the FSR data 
+       if necessary where -1 indicates the key was already added */
+    fsr_id = _FSR_keys_map.insert_and_get_count(fsr_key_hash, NULL);
+    if( fsr_id == -1)
+    {
+      fsr_data volatile* fsr;
+      do{
+        fsr = _FSR_keys_map.at(fsr_key_hash);
+      } while(fsr == NULL);
+      fsr_id = fsr->_fsr_id;
+    }
     else{
-
-        /* Add FSR information to FSR key map and FSR_to vectors */
-      fsr_id = _num_FSRs;
+      
+      /* Add FSR information to FSR key map and FSR_to vectors */
       fsr_data* fsr = new fsr_data;
       fsr->_fsr_id = fsr_id;
+      _FSR_keys_map.at(fsr_key_hash) = fsr;
       Point* point = new Point();
       point->setCoords(coords->getHighestLevel()->getX(), 
                        coords->getHighestLevel()->getY());
+      
+      /* Get the cell that contains coords */
+      Cell* cell = findCellContainingCoords(curr);
       fsr->_point = point;
-      _FSR_keys_map[fsr_key_hash] = *fsr;
-      _FSRs_to_keys.push_back(fsr_key_hash);
-      _FSRs_to_material_IDs.push_back(cell->getFillMaterial()->getId());
-
-      /* If CMFD acceleration is on, add FSR to CMFD cell */
-      if (_cmfd != NULL){
-        int cmfd_cell = _cmfd->findCmfdCell(coords->getHighestLevel());
-        _cmfd->addFSRToCell(cmfd_cell, fsr_id);
-      }
+      fsr->_mat_id = cell->getFillMaterial()->getId();
 
-      /* Increment FSR counter */
-      _num_FSRs++;
+      /* If CMFD acceleration is on, add FSR CMFD cell to FSR data */
+      if (_cmfd != NULL)
+        fsr->_cmfd_cell = _cmfd->findCmfdCell(coords->getHighestLevel());
     }
-
-    /* Release lock */
-    omp_unset_lock(_num_FSRs_lock);
-
   }
+  
   /* If FSR has already been encountered, get the fsr id from map */
-  else
-    fsr_id = _FSR_keys_map.at(fsr_key_hash)._fsr_id;
+  else{
+    fsr_data volatile* fsr;
```

Comment:
I see what you're saying but I'm a bit hesitant to put it in ParallelHashMap because the ``do-while`` loop arises due to an application-specific quirk. And one that we should probably get rid of - namely that we try to add in a value which is an ``fsr_data`` object containing an attribute ``fsr_id`` which refers to the order in which the element was added. I would like to find a way around doing that altogether. But for now we need to do this synchronization in order to make that strange operation thread safe. I think once this is merged in, I will look at further restructuring the code so that we don't need insert_and_get_count and we don't need the ``do-while`` block when checking the ``fsr_id`` but for now I think the code should stay in its current form.